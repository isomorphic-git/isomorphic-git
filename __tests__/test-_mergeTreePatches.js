/* eslint-env node, browser, jasmine */
const { makeFixture } = require('./__helpers__/FixtureFS.js')
// @ts-ignore
const snapshots = require('./__snapshots__/test-_mergeTreePatches.js.snap')
const registerSnapshots = require('./__helpers__/jasmine-snapshots')

const { _diffTree, _mergeTreePatches } = require('isomorphic-git/internal-apis')

describe('_mergeTreePatches', () => {
  beforeAll(() => {
    registerSnapshots(snapshots)
  })

  it('merge two empty TreePatches', async () => {
    // Setup
    const { fs, gitdir } = await makeFixture('test-_diffTree')
    // Test
    const diff1 = await _diffTree({
      fs,
      gitdir,
      before: 'master',
      after: 'master'
    })
    const diff2 = await _diffTree({
      fs,
      gitdir,
      before: 'master',
      after: 'master'
    })
    const { treePatch, hasConflicts } = await _mergeTreePatches({
      treePatches: [diff1, diff2]
    })
    expect(hasConflicts).toBe(false)
    expect(treePatch).toEqual({ basename: '.', index: 0, ops: [] })
  })

  it('merge two conflicting TreePatches', async () => {
    // Setup
    const { fs, gitdir } = await makeFixture('test-_diffTree')
    // Test
    const diff1 = await _diffTree({
      fs,
      gitdir,
      before: 'master',
      after: 'a-file'
    })
    const diff2 = await _diffTree({
      fs,
      gitdir,
      before: 'master',
      after: 'a-folder'
    })
    const { treePatch, hasConflicts } = await _mergeTreePatches({
      treePatches: [diff1, diff2]
    })
    expect(hasConflicts).toBe(true)
    expect(treePatch).toMatchSnapshot()
  })

  it('merge add-files and change-modes', async () => {
    // Setup
    const { fs, gitdir } = await makeFixture('test-_diffTree')
    // Test
    const diff1 = await _diffTree({
      fs,
      gitdir,
      before: 'mainline',
      after: 'add-files'
    })
    const diff2 = await _diffTree({
      fs,
      gitdir,
      before: 'mainline',
      after: 'change-modes'
    })
    const { treePatch, hasConflicts } = await _mergeTreePatches({
      treePatches: [diff1, diff2]
    })
    expect(hasConflicts).toBe(false)
    expect(treePatch).toMatchSnapshot()
  })

  it('merge add-files and remove-files', async () => {
    // Setup
    const { fs, gitdir } = await makeFixture('test-_diffTree')
    // Test
    const diff1 = await _diffTree({
      fs,
      gitdir,
      before: 'mainline',
      after: 'add-files'
    })
    const diff2 = await _diffTree({
      fs,
      gitdir,
      before: 'mainline',
      after: 'remofe-files'
    })
    const { treePatch, hasConflicts } = await _mergeTreePatches({
      treePatches: [diff1, diff2]
    })
    expect(hasConflicts).toBe(false)
    expect(treePatch).toMatchSnapshot()
  })
})
