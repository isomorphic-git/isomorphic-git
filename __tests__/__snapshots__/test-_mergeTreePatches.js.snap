// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_mergeTreePatches merge add-files and change-modes 1`] = `
Object {
  "basename": ".",
  "index": 0,
  "ops": Array [],
  "subOps": Array [
    Object {
      "after": "78981922613b2afb6025042ff6bd878ac1994e85",
      "basename": "a",
      "before": "78981922613b2afb6025042ff6bd878ac1994e85",
      "index": 1,
      "ops": Array [
        "chmod+x",
      ],
    },
    Object {
      "after": "c19d15d2a33fcfd7633460bf08e25eb70a0b75a9",
      "basename": "c",
      "before": "faceb189bb665fe7a631602d904e57a4014187b7",
      "index": 0,
      "ops": Array [],
      "subOps": Array [
        Object {
          "after": "4bcfe98e640c8284511312660fb8709b0afa888e",
          "basename": "d",
          "before": "4bcfe98e640c8284511312660fb8709b0afa888e",
          "index": 1,
          "ops": Array [
            "chmod+x",
          ],
        },
        Object {
          "after": "01058d844a98d293a3b03a8615a34700e4ed2be3",
          "basename": "g",
          "index": 0,
          "ops": Array [
            "write",
          ],
        },
      ],
    },
    Object {
      "after": "6a69f92020f5df77af6e8813ff1232493383b708",
      "basename": "f",
      "index": 0,
      "ops": Array [
        "write",
      ],
    },
  ],
}
`;

exports[`_mergeTreePatches merge add-files and remove-files 1`] = `
Object {
  "basename": ".",
  "index": 0,
  "ops": Array [],
  "subOps": Array [
    Object {
      "basename": "a",
      "before": "78981922613b2afb6025042ff6bd878ac1994e85",
      "index": 1,
      "ops": Array [
        "rm",
      ],
    },
    Object {
      "basename": "b",
      "before": "61780798228d17af2d34fce4cfbdf35556832472",
      "index": 1,
      "ops": Array [
        "rm",
      ],
    },
    Object {
      "basename": "c",
      "before": "faceb189bb665fe7a631602d904e57a4014187b7",
      "index": 1,
      "ops": Array [
        "rmdir",
      ],
      "subOps": Array [
        Object {
          "basename": "d",
          "before": "4bcfe98e640c8284511312660fb8709b0afa888e",
          "index": 1,
          "ops": Array [
            "rm",
          ],
        },
        Object {
          "basename": "e",
          "before": "d905d9da82c97264ab6f4920e20242e088850ce9",
          "index": 1,
          "ops": Array [
            "rm",
          ],
        },
        Object {
          "after": "01058d844a98d293a3b03a8615a34700e4ed2be3",
          "basename": "g",
          "index": 0,
          "ops": Array [
            "write",
          ],
        },
      ],
    },
    Object {
      "after": "6a69f92020f5df77af6e8813ff1232493383b708",
      "basename": "f",
      "index": 0,
      "ops": Array [
        "write",
      ],
    },
  ],
}
`;

exports[`_mergeTreePatches merge two conflicting TreePatches 1`] = `
Object {
  "basename": ".",
  "index": 0,
  "ops": Array [],
  "subOps": Array [
    Object {
      "conflicts": Array [
        Object {
          "after": "78981922613b2afb6025042ff6bd878ac1994e85",
          "basename": "a",
          "index": 0,
          "ops": Array [
            "write",
          ],
        },
        Object {
          "after": "6be660545b31f61a82a87d2b1915f0b88bb9f16f",
          "basename": "a",
          "index": 1,
          "ops": Array [
            "mkdir",
          ],
        },
      ],
      "ops": Array [],
      "subOps": Array [
        Object {
          "after": "61780798228d17af2d34fce4cfbdf35556832472",
          "basename": "b",
          "index": 1,
          "ops": Array [
            "write",
          ],
        },
      ],
    },
  ],
}
`;
